covid19_sfbayarea/data/solano.py

```diff
@@ -0,0 +1,287 @@
+
+    re_keyed = [{TIMESERIES_KEYS[key]: value for key, value in entry.items()}
+                for entry in features]
+
+    # Templates have all data points in the data model
+    # datapoints that are not reported have a value of -1
+    CASES_TEMPLATE = { "date": -1, "cases":-1, "cumul_cases":-1 }
+    DEATHS_TEMPLATE = {"date":-1, "deaths":-1, "cumul_deaths":-1 }
+    TESTS_TEMPLATE = { "date": -1, "tests": -1, "positive": -1, "negative": -1, "pending": -1, "cumul_tests": -1,"cumul_pos": -1, "cumul_neg": -1, "cumul_pend": -1 }
+
+    for entry in re_keyed:
+        #FIXME:
+        # Sonoma county has entries for each day, but some days have "null" for cumulative deaths and cumulative tests.
+        # This means that cumulative counts go something like: 1, null ,null, null, 2, 3 -- on consecutive days.
+        # My best guess is that for deaths and tests they are entering "null" for the cumulative counts on days where the cumulative counts haven't changed from the previous day.
+        # The solution below is just to filter out the cumulative count "null" values (Python None), which results in gaps in the timeseries.
+        # We may want to revisit this and clean up our json to interpolate the values on null-count days until the day the value changes. But that means we're injecting numbers
+        # into the data, and I don't know if we want to do that. But we will be interpolating anyways, in our visualizations, so maybe that's ok.
+
+        # grab cumulative values, to check if they are None
+        cumul_cases = entry["cumul_cases"]
+        cumul_deaths = entry["cumul_deaths"]
+        cumul_tests = entry["cumul_tests"]
+
+        # grab cases and replace None with -1
+        if entry["cases"] is None:
+            entry["cases"] = -1
+
+        if cumul_cases is not None:
+            cases_entry = { k: entry[k] for k in CASES_TEMPLATE if k in entry }
+            series["cases"].append(cases_entry)
+        if cumul_deaths is not None:
+            deaths_entry = {k: entry[k] for k in DEATHS_TEMPLATE if k in entry}
+            series["deaths"].append(deaths_entry)
+        if cumul_tests is not None:
+            tests_entry = {k: entry[k] for k in TESTS_TEMPLATE if k in entry}
+            series["tests"].append(tests_entry)
+
+    out["series"].update(series)
+
+
+def get_notes() -> str:
+    """Scrape notes and disclaimers from dashboard."""
+    # As of 6/5/20, the only disclaimer is "Data update weekdays at 4:30pm"
+    notes = []
+    match = re.compile('[Dd]isclaimers?')
+    driver = get_firefox()
+    driver.implicitly_wait(30)
+    driver.get(dashboard_url)
+    soup = BeautifulSoup(driver.page_source, 'html5lib')
+    has_notes = False
+    text = soup.get_text().splitlines()
+    for text_item in text:
+        if match.search(text_item):
+            notes.append(text_item.strip())
+            has_notes = True
+    if not has_notes:
+        raise(FutureWarning(
```

Comment:
The `sonoma` branch has a custom `FormatError` for issues like this that we should start using as soon as it is merged in

Reply:
Ok, I just added a `data/errors.py` in #84 that includes `FormatError` and `BadRequest`. That could be a place to define custom exceptions. I've left this as a `TODO` for now in this scraper, until we decide on organization for our error classes.

Reply:
FYI I changed this to put `FormatError` in `covid19_sfbayarea/errors.py` since `news` also uses this exception